lastgersync/evmdownloader_fep.go

```diff
@@ -0,0 +1,210 @@
+	l1InfoTreeSync L1InfoTreeQuerier,
+	processor *processor,
+	rh *sync.RetryHandler,
+	blockFinality *big.Int,
+	waitForNewBlocksPeriod time.Duration,
+) (*downloaderFEP, error) {
+	l2GERManager, err := polygonzkevmglobalexitrootv2.NewPolygonzkevmglobalexitrootv2(
+		l2GERAddr, l2Client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize L2 GER manager contract: %w", err)
+	}
+
+	evmDownloader := sync.NewEVMDownloaderImplementation(
+		"lastgersync", l2Client, blockFinality,
+		waitForNewBlocksPeriod, nil, nil,
+		rh, nil)
+
+	return &downloaderFEP{
+		EVMDownloaderImplementation: evmDownloader,
+		l2GERManager:                l2GERManager,
+		l2GERAddr:                   l2GERAddr,
+		l1InfoTreeSync:              l1InfoTreeSync,
+		processor:                   processor,
+		rh:                          rh,
+	}, nil
+}
+
+// RuntimeData returns the runtime data: chainID + addresses to query
+func (d *downloaderFEP) RuntimeData(ctx context.Context) (sync.RuntimeData, error) {
+	chainID, err := d.ChainID(ctx)
+	if err != nil {
+		return sync.RuntimeData{}, err
+	}
+	return sync.RuntimeData{
+		ChainID:   chainID,
+		Addresses: []common.Address{d.l2GERAddr},
+	}, nil
+}
+
+func (d *downloaderFEP) Download(ctx context.Context, fromBlock uint64, downloadedCh chan sync.EVMBlock) {
+	var (
+		attempts            int
+		nextL1InfoTreeIndex uint32
+		err                 error
+	)
+
+	// Determine the next index to start fetching GERs
+	for {
+		latestL1InfoTreeIndex, err := d.processor.getLatestL1InfoTreeIndex()
+		if errors.Is(err, db.ErrNotFound) {
+			nextL1InfoTreeIndex = 0
+		} else if err != nil {
+			log.Errorf("error getting latest l1 info tree index: %v", err)
+			attempts++
+			d.rh.Handle("getLatestL1InfoTreeIndex", attempts)
+
+			continue
+		}
+		if latestL1InfoTreeIndex > 0 {
+			nextL1InfoTreeIndex = latestL1InfoTreeIndex + 1
+		}
+		break
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debug("aborting the lastgersync downloader...")
+			close(downloadedCh)
+
+			return
+		default:
+		}
+
+		// Wait for new blocks before processing
+		fromBlock = d.WaitForNewBlocks(ctx, fromBlock)
+
+		// Fetch GERs from the determined index
+		attempts = 0
+		var gers []*GlobalExitRootInfo
+		for {
+			gers, err = d.getGERsFromIndex(ctx, nextL1InfoTreeIndex)
+			if err != nil {
+				log.Errorf("error getting GERs: %v", err)
+				attempts++
+				d.rh.Handle("getGERsFromIndex", attempts)
+
+				continue
+			}
+
+			break
+		}
+
+		header, isCanceled := d.GetBlockHeader(ctx, fromBlock)
+		if isCanceled {
+			return
+		}
+
+		block := &sync.EVMBlock{
+			EVMBlockHeader: sync.EVMBlockHeader{
+				Num:        header.Num,
+				Hash:       header.Hash,
+				ParentHash: header.ParentHash,
+				Timestamp:  header.Timestamp,
+			},
+		}
+		// Set the greatest GER injected from retrieved GERs
+		d.populateGreatestInjectedGER(block, gers)
+
+		// Update nextIndex based on the last injected GER info
+		if len(block.Events) > 0 {
```

Comment:
I still don't get this logic. Isn't this why we wanted to separate the two sync methods?

Reply:
Great point, it was a leftover that accidentally remained :raised_hands: 

Fixed https://github.com/agglayer/aggkit/pull/357/commits/aa160bd90b79c2074252212341cfe547d2598eb3